pixeltable/utils/documents.py

```diff
@@ -9,31 +9,56 @@ class DocumentHandle:
     format: ts.DocumentType.DocumentFormat
     bs_doc: Optional['bs4.BeautifulSoup'] = None
     md_ast: Optional[Dict] = None
+    pdf_doc: Optional['fitz.Document'] = None
 
-
-def get_document_handle(s: str) -> Optional[DocumentHandle]:
-    bs_doc = get_html_handle(s)
-    if bs_doc is not None:
-        return DocumentHandle(format=ts.DocumentType.DocumentFormat.HTML, bs_doc=bs_doc)
-    md_ast = get_markdown_handle(s)
+def get_document_handle(path: str) -> Optional[DocumentHandle]:
+    # NB: try pdf first, because correct PDF must be opened in binary mode,
+    # markdown and HTML would throw an error if they open pdf.
+    pdf_doc = get_pdf_handle(path)
+    if pdf_doc is not None:
+        return DocumentHandle(format=ts.DocumentType.DocumentFormat.PDF, pdf_doc=pdf_doc)
+    md_ast = get_markdown_handle(path)
     if md_ast is not None:
         return DocumentHandle(format=ts.DocumentType.DocumentFormat.MD, md_ast=md_ast)
-    return None
+    # NB: get_html_handle must be last, because it will return a handle for md files as well,
+    # and we want to prefer the md handle.
+    bs_doc = get_html_handle(path)
+    if bs_doc is not None:
+        return DocumentHandle(format=ts.DocumentType.DocumentFormat.HTML, bs_doc=bs_doc)
 
-def get_html_handle(s: str) -> Optional['bs4.BeautifulSoup']:
+    return pdf_doc
+
+def get_html_handle(path: str) -> Optional['bs4.BeautifulSoup']:
     import bs4
-    try:
-        doc = bs4.BeautifulSoup(s, 'html.parser')
-    except Exception as e:
-        return None
-    if doc.find() is None:
-        return None
-    return doc
+    with open(path, 'r', encoding='utf8') as file:
+        s = file.read()
+        try:
+            doc = bs4.BeautifulSoup(s, 'html.parser')
+        except Exception:
+            return None
+        if doc.find() is None:
+            return None
+        return doc
 
-def get_markdown_handle(s: str) -> Optional[Dict]:
+def get_markdown_handle(path: str) -> Optional[Dict]:
     import mistune
+    with open(path, 'r', encoding='utf8') as file:
+        s = file.read()
```

Comment:
If we have to try both markdown and HTML, we end up reading the entire file twice into a string. That's inefficient? If we're going to read the file into memory anyway, can't we do it once and then try both MD and html against the string?